### PR TITLE
rpc rebase: Try-runtime Revamp and Facelift

### PR DESCRIPTION
**Pull request by lexnv**
This PR merges the `origin/master` into https://github.com/paritytech/substrate/pull/12537.

The main focus of this PR is to handle the new batch request API exposed by the jsonrpsee
in `utils/frame/remote-externalities/src/lib.rs`.

Jsonrpsee changed the API of batch requests and failures are reported via `Err(ErrorObject<'static>)`.

To reflect this change, a new inter-thread messaging is added `BatchFailed`.
Whenever the `BatchFailed` message is encountered, we stop the main waiting loop
and ensure that threads are joined before returning.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2154483

File: utils/frame/remote-externalities/src/lib.rs [hunk 1]
```diff
@@ -427,57 +433,75 @@ where
...
-						.map(|key| ("state_getStorage", rpc_params![key, at]))
-						.collect::<Vec<_>>();
+					let mut batch = BatchRequestBuilder::new();
 
-					let values = rt
+					for key in chunk_keys.iter() {
+						batch
+							.insert("state_getStorage", rpc_params![key, at])
+							.map_err(|_| "Invalid batch params")
+							.unwrap();
```
Comment (kianenigma) [inline utils/frame/remote-externalities/src/lib.rs:444]: return err?

File: utils/frame/remote-externalities/src/lib.rs [hunk 2]
```diff
@@ -1087,88 +1138,19 @@ mod tests {
 #[cfg(all(test, feature = "remote-test"))]
 mod remote_tests {
 	use super::test_prelude::*;
-	use std::os::unix::fs::MetadataExt;
```
Comment (kianenigma) [inline utils/frame/remote-externalities/src/lib.rs:1090]: why all of these code removals? I think you branched off form the wrong branch. Start from the latest code in my base branch, and merge master please.

**Review by kianenigma — CHANGES_REQUESTED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2154946

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2155321

**Comment by lexnv** [comment]
bot merge

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2157561

**Pull request merged**